cardano-cli/test/cardano-cli-test/Test/Cli/Parser.hs

```diff
@@ -0,0 +1,105 @@
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Test.Cli.Parser
+  ( hprop_integral_reader
+  , hprop_integral_pair_reader_positive
+  , hprop_integral_pair_reader_negative
+  )
+where
+
+import           Cardano.CLI.EraBased.Options.Common (integralParsecParser,
+                   pairIntegralParsecParser)
+
+import           Data.Bits (Bits)
+import           Data.Data (Proxy (..), Typeable)
+import           Data.Either (isLeft, isRight)
+import           Data.Word (Word16)
+import qualified Text.Parsec as Parsec
+
+import           Hedgehog (Gen, Property, assert, property)
+import           Hedgehog.Extras (propertyOnce)
+import qualified Hedgehog.Gen as Gen
+import           Hedgehog.Internal.Property (forAll)
+import qualified Hedgehog.Range as Range
+
+-- | Execute me with:
+-- @cabal test cardano-cli-test --test-options '-p "/integral reader/"'@
+hprop_integral_reader :: Property
+hprop_integral_reader = propertyOnce $ do
+  assert $ isRight $ parse @Word "0"
+  assert $ isRight $ parse @Word "42"
+  assert $ isLeft $ parse @Word "-1"
+  assert $ isLeft $ parse @Word "18446744073709551616"
+  assert $ isLeft $ parse @Word "-1987090"
+
+  assert $ isRight $ parse @Word16 "0"
+  assert $ isRight $ parse @Word16 "42"
+  assert $ isLeft $ parse @Word16 "-1"
+  assert $ isLeft $ parse @Word16 "65536"
+  assert $ isLeft $ parse @Word16 "298709870987"
+  assert $ isLeft $ parse @Word16 "-1987090"
+ where
+  parse :: (Typeable a, Integral a, Bits a) => String -> Either String a
+  parse s =
+    case Parsec.runParser integralParsecParser () "" s of
+      Left parsecError -> Left $ show parsecError
+      Right x -> Right x
+
+-- | Execute me with:
+-- @cabal test cardano-cli-test --test-options '-p "/integral pair reader positive/"'@
+hprop_integral_pair_reader_positive :: Property
+hprop_integral_pair_reader_positive = property $ do
+  validArbitraryTuple <- forAll $ genNumberTuple (Proxy :: Proxy Word)
+  assert $ isRight $ parse @Word validArbitraryTuple
+
+  assert $ isLeft $ parse @Word "(0, 0, 0)"
```

Comment:
You can write a generator that injects a random character (or characters) into the string generated by `genNumberTuple`. This would provide better coverage.